<a href="https://colab.research.google.com/github/jarodchristiansen/PythonProjects/blob/master/Logistic_Regression_With_Statsmodels.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install --upgrade -q gspread
!pip install gspread-dataframe

These requirements should normally be satisfied, they are dependencies for gspread to pull the dataframe

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn
seaborn.set()
import gspread
from gspread_dataframe import get_as_dataframe, set_with_dataframe

Import necessary libraries to create dataframe, regression, and formatting of regression (plotting) Gspread creates dataframe from sheets file

In [ ]:
from google.colab import auth
auth.authenticate_user()
from oauth2client.client import GoogleCredentials
gc = gspread.authorize(GoogleCredentials.get_application_default())

This will create authorization through google sheets (follow link that appears and copy authorization code)

In [ ]:
sheet1 = gc.open_by_url('https://docs.google.com/spreadsheets/d/1mD8g8oX4Vq7EK-3OCHmFHtjLcs17yyzCSoFSmeNcDms/edit#gid=0')

sheet1 is entire csv file URL is sheets file URL

In [ ]:
ws = sheet1.worksheet('SAT')

This is the tab in csv file

In [ ]:
data = get_as_dataframe(ws)

titles the dataframe extracted from sheets file

In [ ]:
data

SAT Admitted  Unnamed: 2  ...  Unnamed: 23  Unnamed: 24  Unnamed: 25
0    1363.0       No         NaN  ...          NaN          NaN          NaN
1    1792.0      Yes         NaN  ...          NaN          NaN          NaN
2    1954.0      Yes         NaN  ...          NaN          NaN          NaN
3    1653.0       No         NaN  ...          NaN          NaN          NaN
4    1593.0       No         NaN  ...          NaN          NaN          NaN
..      ...      ...         ...  ...          ...          ...          ...
994     NaN      NaN         NaN  ...          NaN          NaN          NaN
995     NaN      NaN         NaN  ...          NaN          NaN          NaN
996     NaN      NaN         NaN  ...          NaN          NaN          NaN
997     NaN      NaN         NaN  ...          NaN          NaN          NaN
998     NaN      NaN         NaN  ...          NaN          NaN          NaN

[999 rows x 26 columns]

This is calling the dataframe and representing from Pandas. NaN cells are empty values

In [ ]:
data.dropna(axis=0, how='all', inplace=True)

In [ ]:
data.dropna(axis=1, how='all', inplace=True)

In [ ]:
data

SAT Admitted
0    1363.0       No
1    1792.0      Yes
2    1954.0      Yes
3    1653.0       No
4    1593.0       No
..      ...      ...
163  1722.0      Yes
164  1750.0      Yes
165  1555.0       No
166  1524.0       No
167  1461.0       No

[168 rows x 2 columns]

In [ ]:
data.describe(include='all')

SAT Admitted
count    168.000000      168
unique          NaN        2
top             NaN      Yes
freq            NaN       94
mean    1695.273810      NaN
std      183.019017      NaN
min     1334.000000      NaN
25%     1547.500000      NaN
50%     1691.500000      NaN
75%     1844.500000      NaN
max     2050.000000      NaN

In [ ]:
data['Admitted'] = data['Admitted'].map({'Yes':1, 'No':0})

In [ ]:
data

SAT  Admitted
0    1363.0         0
1    1792.0         1
2    1954.0         1
3    1653.0         0
4    1593.0         0
..      ...       ...
163  1722.0         1
164  1750.0         1
165  1555.0         0
166  1524.0         0
167  1461.0         0

[168 rows x 2 columns]

This is creating dummy variables regarding attendance

In [ ]:
y = data['Admitted']
x1 = data['SAT']

In [ ]:
x = sm.add_constant(x1)

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [ ]:
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()

Optimization terminated successfully.
         Current function value: 0.137766
         Iterations 10


reg_log = sm.Logit() is performing a Logit regression. The next line of results_log = reg_log.fit() is assigning a variable equal to the line of regression acorrding to variables input into the Logit regression function. The number of iterations has to be under 35 for the machine learning algorithm to have learned properly. 

In [ ]:
results_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Sun, 16 Feb 2020   Pseudo R-squ.:                  0.7992
Time:                        02:54:52   Log-Likelihood:                -23.145
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.805e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -69.9128     15.737     -4.443      0.000    -100.756     -39.070
SAT            0.0420      0.009      4.454      0.000       0.024       0.060
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

Method : MLE = Maximum Likelihood estimation 

Likelihood function:
a function which estimates how likely it is that the model
at hand describes the real underlying relationship of variables


the bigger the likelihood funciton. the higher the probabilty
that our model is correct

MLE tries to maximize the likelihood function


The computer is going through different values, until it finds
a model, for which the likelihood is the highest.

When it can no longer improve it, it will just stop the optimization

(Which is standard in machine learning processes.)

logit model 
log(pi/1-pi) = -69.91+0.042 * SAT
or log (p(X)/1-p(X)) or the probability of an event occurring
divided by the probability of it not occurring

ex: log(odds2) = -69.91+0042*SAT2
ex: log(odds1) = -69.91+0042*SAT1

or log(odds2) - log(odds1) = 0.042(SAT2 - SAT1)

or log(odds2/odds1) = 0.042(SAT2 - SAT1) maybe the most efficient format

or log(odds2/odds1) = 0.042


difference of 1 unit of SAT: 

log(odds2/odds1) = 0.042


e_log(odds2/odds1) = e0.042 or
odds2/odds1 = 1.042
this indicates that odds2 = 104.2%*odds1 
or when SAT score increseases by 1 the admittance increases
by 4.2% 


unitchange in odds = exponent of a coefficient


Difference of 10 units of SAT example
log(odds2/odds1) = 0.042 8 10